### ISSUE TITLE
executeSql not called in transaction

### ISSUE THREAD

**Issue by thsantac** [brody2consult/cordova-sqlite-storage#514]
Hi,

I am getting crazy...

I have spent a whole day trying to figure out what's wrong in this function, so any help will be greatly appreciated.

I test this code in xCode and its iPhone Emulator. I open the database with success, populate a table with success and retrieve 175 news via Ajax call and put the JSON I get in an array (global_news_data).

This the function that's driving me crazy (in my DBService class), it just shows the second console.log 175 times ("SELECT news_id=xxxxx") with good news_ids and that's all:

```
    this.saveNews = function() {

        var deferred             = $.Deferred();

        this.db.transaction(

            function (tx) {

                console.log("saveNews - global_news_data.rows.length="+global_news_data.rows.length);
                for(var i = 0; i < global_news_data.rows.length; i++)
                {
                    if(global_news_data.rows[i].ID != null) {

                        console.log("SELECT news_id="+global_news_data.rows[i].ID);

                        tx.executeSql("SELECT * FROM news WHERE news_id = ?", [global_news_data.rows[i].ID], 
                            function(tx, res) {
                                console.log("SELECT news_id="+global_news_data.rows[i].ID+" OK !!!");
                                if(res.rows.length==0) {
                                    tx.executeSql("INSERT INTO news (news_id, news_post_title, news_post_date_gmt, news_post_date_long, news_post_content, news_comment_count, news_views, news_image, news_author) VALUES (?, ?, ?, ?, ?, ?, ?, ?)", [global_news_data.rows[i].ID, global_news_data.rows[i].post_title, global_news_data.rows[i].post_date_gmt, global_news_data.rows[i].post_date_long, global_news_data.rows[i].post_content, global_news_data.rows[i].comment_count, global_news_data.rows[i].views, global_news_data.rows[i].img, global_news_data.rows[i].display_name], 
                                        function(tx, res) {
                                            console.log("news inserted !!!");
                                        },
                                        function (tx, error) {
                                            console.log('news INSERT error : ' + error.message);
                                            return true;
                                        }
                                    );
                                }
                                else {
                                    tx.executeSql("UPDATE news SET news_comment_count = ?, news_views = ? WHERE news_id = ?", [global_news_data.rows[i].comment_count, global_news_data.rows[i].views, global_news_data.rows[i].ID], 
                                        function(tx, res) {
                                            console.log("news updated !!!");
                                        },
                                        function (tx, error) {
                                            console.log('news UPDATE error : ' + error.message);
                                            return true;
                                        }
                                    );
                                }
                            },
                            function (tx, error) {
                                console.log('news SELECT error : ' + error.message);
                                return true;
                            }
                        );
                    }
                }
            },
            function (error) {
                deferred.reject("saveNews transaction Error: "+error.code+", msg:"+error.message);
            },
            function () {
                console.log('saveNews transaction success');
                deferred.resolve();
            }
        );
        return deferred.promise();
    }

```

None of the transaction callbacks are called. I just see "SELECT news_id=xxx" 175 times because I have 175 news to update. It seems the first tx.executeSql in not called.

Thanks a lot.

**Comment by brody2consult** [comment]
Unfortunately I find the code a bit hard to follow and cannot be sure where all of your data is coming from.

It appears that you are using a ResultSet object from another transaction. While this should be OK I really find it messy and hard to follow.

You should be able to use JOINs to get your data in one shot, and you can also INSERT with data from a SELECT statement.

I suggest you post this as a question about SQLite in general on Stack Overflow and see if you get any answers.

**Comment by thsantac** [comment]
Thanks for the reply.

I just added a console.dir just after the transaction begins and this is what I get : 

```
`2016-07-08 12:28:30.799 SNCTA_Mobile_App[30184:2957746] {"db":{"openargs":{"name":"sncta.db","location":"default","dblocation":"docs"},"dbname":"sncta.db"},"txlock":true,"readOnly":false,"executes":[{"success":null,"qid":0,"sql":"BEGIN","params":[]}]}
2016-07-08 12:28:30.800 SNCTA_Mobile_App[30184:2957746] saveNews - global_news_data.rows.length=175
2016-07-08 12:28:30.801 SNCTA_Mobile_App[30184:2957746] SELECT news_id=10346
2016-07-08 12:28:30.805 SNCTA_Mobile_App[30184:2957746] SELECT news_id=10199
2016-07-08 12:28:30.805 SNCTA_Mobile_App[30184:2957746] SELECT news_id=10186
```

I put the console.dir here : 

```
        this.db.transaction(

            function (tx) {
                **console.dir(tx);**
                console.log("saveNews - global_news_data.rows.length="+global_news_data.rows.length);
                for(var i = 0; i < global_news_data.rows.length; i++)
                {
                    if(global_news_data.rows[i].ID != null) {

                        console.log("SELECT news_id="+global_news_data.rows[i].ID);

                        tx.executeSql("SELECT * FROM news WHERE news_id = ?", [global_news_data.rows[i].ID], 

```

Is the tx object OK for you ?

Thank you.

**Comment by brody2consult** [comment]
Unfortunately I really cannot follow what is going wrong. I think it would be best if you can post a Cordova project that demonstrates the issue, with no platforms or plugins included. (You can do `cordova plugin add` with the `--save` flag to add each plugin to your `config.xml` file.)

**Comment by thsantac** [comment]
I just found the problem: if I remove the for() instruction, declare var i = 0 and just try to update the first news it works !!! 

```
                //for(var i = 0; i < global_news_data.rows.length; i++)
                //{
                    //if(global_news_data.rows[i].ID != null) {

                        var i = 0;
                        console.log("SELECT news_id="+global_news_data.rows[i].ID);

                        tx.executeSql("SELECT * FROM news WHERE news_id = ?", [global_news_data.rows[i].ID], 

```

This time the tx.executeSql is called. So the issue is the loop but I don't understand why...
Do you have any ideas?

Thanks

**Comment by brody2consult** [comment]
Unfortunately I really cannot say for sure. It is definitely not allowed to use a transaction object once a transaction has finished but I don't think that should be the case here.

You may want to try a while loop with i=0 only, then with i=0 and 1 and see what happens in these cases.

**Comment by thsantac** [comment]
OK I found a workaround :-) I post it so it may help someone else. I have put the inner transaction code in a function that I wait for it to be completely executed : 

```
        this.db.transaction(
            function (tx) {
                for(var i = 0; i < global_news_data.rows.length; i++)
                {
                    if(global_news_data.rows[i].ID != null) {
                        var news_id                 = global_news_data.rows[i].ID;
                        var data                    = global_news_data.rows[i];

                        saveOneNews(tx, data).done(function() {
                            console.log("News updated :"+news_id);
                        });

                    }
                }
            },
            function (error) {
                deferred.reject("saveNews transaction Error: "+error.code+", msg:"+error.message);
            },
            function () {
                console.log('saveNews transaction success');
                deferred.resolve();
            }
        );

```

And the new function:

```
    var saveOneNews = function(tx, data) {

        var deferred             = $.Deferred();

        if(data.ID != null) {

            var news_id                 = data.ID;
            console.log("SELECT news_id="+news_id);

            tx.executeSql("SELECT * FROM news WHERE news_id = ?", [news_id], 
                function(tx, res) {
                    console.log("SELECT news_id="+news_id+" OK !!!");
                    if(res.rows.length==0) {
                        tx.executeSql("INSERT INTO news (news_id, news_post_title, news_post_date_gmt, news_post_date_long, news_post_content, news_comment_count, news_views, news_image, news_author) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)", [data.ID, data.post_title, data.post_date_gmt, data.post_date_long, data.post_content, data.comment_count, data.views, data.img, data.display_name], 
                            function(tx, res) {
                                console.log("news inserted !!!");
                                deferred.resolve();
                            },
                            function (tx, error) {
                                console.log('news INSERT error : ' + error.message);
                                deferred.resolve();
                                return true;
                            }
                        );
                    }
                    else {
                        tx.executeSql("UPDATE news SET news_comment_count = ?, news_views = ? WHERE news_id = ?", [data.comment_count, data.views, data.ID], 
                            function(tx, res) {
                                deferred.resolve();
                                console.log("news updated !!!");
                            },
                            function (tx, error) {
                                console.log('news UPDATE error : ' + error.message);
                                deferred.resolve();
                                return true;
                            }
                        );
                    }
                },
                function (tx, error) {
                    console.log('news SELECT error : ' + error.message);
                    deferred.resolve();
                    return true;
                }
            );
        }
        return deferred.promise();
    }


```

And now it works :-)

Thanks again.

**Comment by brody2consult** [comment]
Thanks for sharing.
